services/accumulo-utils/src/main/java/datawave/security/util/AuthorizationsMinimizer.java

```diff
@@ -0,0 +1,68 @@
+     * entities and using that set), is the fact that Accumulo visibility expressions may include disjunctions. Consider the following visibility expression.
+     * <p>
+     * {@code (A & B & ORG1) | (A & C & ORG2)}
+     * <p>
+     * If the incoming user has the Accumulo authorizations A, B, and ORG1 and the user is coming through a server with the Accumulo authorizations A, C, and
+     * ORG2, then the server is allowed to see data protected with this expression and the user is also allowed, so the data should be returned with those two
+     * calling entities. If we simply computed an intersection of the authorizations that set would be A. That set of authorizations does not match the
+     * visibility expression and data that should be returned would not be.
+     * <p>
+     * Given that we must apply multiple visibility filters in order to properly return data for a calling entity chain, it is desirable to minimize the number
+     * of visibility filters used, if possible. If a set of authorizations is a superset of any other set of authorizations, then it does not need to be tested.
+     * As long as we are not changing any of the authorizations sets, we know we will never allow any data to be returned that would otherwise not be. By
+     * removing authorizations sets, we might only prevent the return of data that should otherwise be returned. Consider a given set of authorizations that is
+     * a subset of another. If the subset passes a visibility expression, then the superset will too since it contains at least the same authorizations, so
+     * there is no need to test against it. If the subset does not pass, then the data is not visible to the chain anyway, so there is no point in testing the
+     * superset.
+     *
+     * @param authorizations
+     *            the list of Authorizations for each entity in the call chain
+     * @return a minimized set of Authorizations that allows visibility of exactly the same data as {@code authorizations}
+     */
+    public static Collection<Authorizations> minimize(Collection<Authorizations> authorizations) {
+        if (authorizations.size() > 1) {
+            // Convert collection of Authorizations into a collection of String sets (the individual authorizations).
+            // Since we are adding to a LinkedHashSet, this will de-dupe any duplicate authorization sets.
+            final LinkedHashSet<TreeSet<String>> allAuths = authorizations.stream()
+                            .map(a -> a.getAuthorizations().stream().map(String::new).collect(Collectors.toCollection(TreeSet::new)))
+                            .collect(Collectors.toCollection(LinkedHashSet::new));
+            
+            // Go through the authorizations sets and remove any that are supersets of any other.
+            for (Iterator<TreeSet<String>> it = allAuths.iterator(); it.hasNext(); /* empty */) {
```

Comment:
Do we want to check for a threshold wherein it is no longer worth doing an n^2 operation?

Reply:
Check size before doing containsall

Reply:
HashSet vs TreeSet?

Reply:
I'm not sure on the threshold. I'm kind of thinking that a threshold big enough to worry about the n^2 setup would mean we'll have so many authorization sets that running each entry through that many visibility filters would be so slow as to be unusable. I feel that that's one of those "Doctor, it hurts when I do that." scenarios.
Will make the other changes.